File: src/lib/coverageReport.ts
import { readStateTileJson, validateStateTileManifest } from './stateTiles'
import type { StateTileManifest } from './stateTiles'

const TIME_SCALE = 'TDB seconds past J2000'
const MAX_SUMMARY_BYTES = 64 * 1024
type ObjectValue = Record<string, unknown>
function object(value: unknown): ObjectValue {
  if (!value || typeof value !== 'object' || Array.isArray(value)) throw new Error('Invalid coverage object')
  return value as ObjectValue
}
function count(value: unknown): number {
  if (typeof value !== 'number' || !Number.isSafeInteger(value) || value < 0) throw new Error('Invalid coverage count')
  return value
}
function finite(value: unknown): number {
  if (typeof value !== 'number' || !Number.isFinite(value)) throw new Error('Invalid coverage epoch')
  return value
}
function hash(value: unknown): string {
  if (typeof value !== 'string' || !/^[a-f0-9]{64}$/.test(value)) throw new Error('Invalid coverage hash')
  return value
}
function coverageText(value: unknown, field: string): string {
  if (typeof value !== 'string' || !value || value.length > 512 || [...value].some(character => character.charCodeAt(0) < 32 || character.charCodeAt(0) === 127)) throw new Error(`Invalid coverage ${field}`)
  return value
}

/** An audit of a pinned source population, never the current rendered frame. */
export function validateCoverageReport(raw: unknown, manifest: StateTileManifest) {
  const value = object(raw)
  if (value.apiVersion !== 'solar.api/v1' || value.purpose !== 'source-identity-and-dependency-window-audit'
    || value.profile !== 'full' || value.sourceBytesVerified !== true
    || value.timeScale !== TIME_SCALE || value.frame !== 'ECLIPJ2000') throw new Error('Invalid coverage contract')
  const catalogManifestSha256 = hash(value.catalogManifestSha256)
  const inventoryManifestSha256 = hash(value.inventoryManifestSha256)
  if (value.catalogVersion !== manifest.catalogVersion || catalogManifestSha256 !== manifest.catalogManifestSha256
    || inventoryManifestSha256 !== manifest.inventoryManifestSha256) throw new Error('Coverage dataset mismatch')
  const inputCounts = object(value.counts)
  const counts = {
    sourceRecords: count(inputCounts.sourceRecords),
    mappedSourceRecords: count(inputCounts.mappedSourceRecords),
    unresolvedSourceRecords: count(inputCounts.unresolvedSourceRecords),
    explicitNaifTargets: count(inputCounts.explicitNaifTargets),
    availableTargetsAtAuditEpoch: count(inputCounts.availableTargetsAtAuditEpoch),
  }
  if (counts.mappedSourceRecords > counts.sourceRecords || counts.unresolvedSourceRecords !== counts.sourceRecords - counts.mappedSourceRecords
    || counts.explicitNaifTargets > counts.mappedSourceRecords || counts.availableTargetsAtAuditEpoch > counts.explicitNaifTargets) throw new Error('Inconsistent coverage counts')
  const inputWindows = object(value.windowCounts)
  const windowCounts = {
    dependencyCoveredTargets: count(inputWindows.dependencyCoveredTargets),
    targetsWithDependencyGaps: count(inputWindows.targetsWithDependencyGaps),
    numericallyCertifiedWholeWindowTargets: null,
  }
  if (inputWindows.numericallyCertifiedWholeWindowTargets !== null
    || windowCounts.dependencyCoveredTargets > counts.explicitNaifTargets
    || windowCounts.targetsWithDependencyGaps !== counts.explicitNaifTargets - windowCounts.dependencyCoveredTargets) throw new Error('Invalid coverage window counts')
  const reasons = Object.entries(object(value.unresolvedReasons))
  if (reasons.length > 128) throw new Error('Too many coverage reasons')
  const unresolvedReasons = reasons.map(([reason, total]) => {
    if (!/^[a-z0-9][a-z0-9-]{0,127}$/.test(reason)) throw new Error('Invalid coverage reason')
    return { reason, count: count(total) }
  }).sort((a, b) => a.reason.localeCompare(b.reason))
  let remaining = counts.unresolvedSourceRecords
  for (const entry of unresolvedReasons) {
    if (entry.count > remaining) throw new Error('Inconsistent coverage reasons')
    remaining -= entry.count
  }
  if (remaining !== 0) throw new Error('Incomplete coverage reasons')
  const rawCoverage = value.coverage
  if (!Array.isArray(rawCoverage) || rawCoverage.length < 1 || rawCoverage.length > 256) throw new Error('Invalid coverage status buckets')
  const coverage = rawCoverage.map(raw => {
    const bucket = object(raw)
    const datasetVersion = coverageText(bucket.datasetVersion, 'dataset')
    const source = coverageText(bucket.source, 'source')
    const model = coverageText(bucket.model, 'model')
    const frame = coverageText(bucket.frame, 'frame')
    const auditEt = finite(bucket.auditEt)
    const exact = count(bucket.exact), approximate = count(bucket.approximate), missing = count(bucket.missing)
    if (datasetVersion !== manifest.catalogVersion || frame !== 'ECLIPJ2000' || auditEt !== finite(value.auditEt)) throw new Error('Invalid coverage status bucket')
    return { datasetVersion, source, model, auditEt, frame, exact, approximate, missing }
  })
  const coverageKeys = new Set(coverage.map(entry => JSON.stringify([entry.datasetVersion, entry.source, entry.model, entry.auditEt, entry.frame])))
  if (coverageKeys.size !== coverage.length) throw new Error('Repeated coverage status bucket')
  const coverageTotal = coverage.reduce((sum, entry) => sum + entry.exact + entry.approximate + entry.missing, 0)
  if (!Number.isSafeInteger(coverageTotal) || coverageTotal !== counts.sourceRecords) throw new Error('Coverage status buckets do not reconcile')
  const inputWindow = object(value.requestedWindow)
  const requestedWindow = { startEt: finite(inputWindow.startEt), endEt: finite(inputWindow.endEt), timeScale: TIME_SCALE }
  if (inputWindow.timeScale !== TIME_SCALE || requestedWindow.startEt > requestedWindow.endEt) throw new Error('Invalid coverage window')
  // The independent audit epoch need not be inside the requested window.
  return {
    catalogVersion: manifest.catalogVersion, catalogManifestSha256, inventoryManifestSha256,
    reportSha256: hash(value.reportSha256), sourceSnapshotSha256: hash(value.sourceSnapshotSha256),
    identityMappingSha256: hash(value.identityMappingSha256), satelliteCatalogSha256: hash(value.satelliteCatalogSha256),
    auditEt: finite(value.auditEt), timeScale: TIME_SCALE, frame: 'ECLIPJ2000',
    requestedWindow, counts, coverage, windowCounts, unresolvedReasons,
  }
}

export type CoverageReport = ReturnType<typeof validateCoverageReport>
export class CoverageUnavailableError extends Error {}

export async function loadCoverageReport(baseUrl: string | null, profile: 'full' | 'preview', signal: AbortSignal): Promise<CoverageReport> {
  // Product restriction is enforced before any network request, even if a
  // preview build accidentally receives a full-backend environment variable.
  if (profile !== 'full' || !baseUrl?.trim()) throw new CoverageUnavailableError('Coverage requires a configured full backend')
  const base = baseUrl.trim().replace(/\/+$/, '')
  const manifest = validateStateTileManifest(await readStateTileJson(await fetch(`${base}/v1/catalog/manifest`, { signal, cache: 'no-store' }), 'Coverage manifest'))
  const response = await fetch(`${base}/v1/coverage`, { signal, cache: 'no-store' })
  if (response.status === 404) { await response.body?.cancel(); throw new CoverageUnavailableError('Coverage report is not configured') }
  const length = Number(response.headers.get('content-length'))
  if (!response.headers.has('content-length') || !Number.isSafeInteger(length) || length < 1 || length > MAX_SUMMARY_BYTES) {
    await response.body?.cancel()
    throw new Error('Coverage summary exceeds its size contract')
  }
  return validateCoverageReport(await readStateTileJson(response, 'Coverage summary'), manifest)
}

File: android/app/src/main/java/io/github/dajiaohuang/solaratlas/CoverageReport.java
package io.github.dajiaohuang.solaratlas;

import java.util.Collections;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.SortedMap;
import java.util.TreeMap;

/** Pinned source-population evidence, never a live rendered-frame count. */
public final class CoverageReport {
    public static final String TIME_SCALE = "TDB seconds past J2000";
    public static final String FRAME = "ECLIPJ2000";
    public final String catalogVersion, catalogHash, inventoryHash, reportHash, sourceHash, mappingHash, satelliteHash;
    public final long sourceRecords, mappedRecords, unresolvedRecords, explicitTargets, availableTargets, dependencyCovered, dependencyGaps;
    public final double auditEt, windowStartEt, windowEndEt;
    public final SortedMap<String, Long> unresolvedReasons;
    public final List<CoverageBucket> coverage;

    public static final class CoverageBucket {
        public final String datasetVersion, source, model, frame;
        public final long exact, approximate, missing;
        public final double auditEt;
        private CoverageBucket(Map<String, Object> value, String catalogVersion, double reportAuditEt) throws StateTileDecoder.ProtocolException {
            datasetVersion = text(value.get("datasetVersion"), false, 512);
            source = text(value.get("source"), false, 512);
            model = text(value.get("model"), false, 512);
            frame = text(value.get("frame"), false, 512);
            auditEt = finite(value.get("auditEt"));
            exact = count(value.get("exact")); approximate = count(value.get("approximate")); missing = count(value.get("missing"));
            require(datasetVersion.equals(catalogVersion) && FRAME.equals(frame) && auditEt == reportAuditEt
                    && sum(exact, approximate, missing) != null, "invalid coverage status bucket");
        }
    }

    private CoverageReport(Map<String, Object> value, Map<String, Object> manifest) throws StateTileDecoder.ProtocolException {
        require("solar.api/v1".equals(manifest.get("apiVersion")), "manifest API mismatch");
        catalogVersion = string(manifest, "catalogVersion");
        require(!catalogVersion.trim().isEmpty(), "manifest catalog version missing");
        catalogHash = hash(manifest, "catalogManifestSha256");
        inventoryHash = hash(manifest, "inventoryManifestSha256");
        require("solar.api/v1".equals(value.get("apiVersion"))
                && "source-identity-and-dependency-window-audit".equals(value.get("purpose"))
                && "full".equals(value.get("profile")) && Boolean.TRUE.equals(value.get("sourceBytesVerified"))
                && TIME_SCALE.equals(value.get("timeScale")) && FRAME.equals(value.get("frame")), "coverage contract mismatch");
        require(catalogVersion.equals(value.get("catalogVersion"))
                && catalogHash.equals(hash(value, "catalogManifestSha256"))
                && inventoryHash.equals(hash(value, "inventoryManifestSha256")), "coverage dataset mismatch");
        reportHash = hash(value, "reportSha256"); sourceHash = hash(value, "sourceSnapshotSha256");
        mappingHash = hash(value, "identityMappingSha256"); satelliteHash = hash(value, "satelliteCatalogSha256");
        Map<String, Object> counts = object(value.get("counts"));
        sourceRecords = count(counts.get("sourceRecords")); mappedRecords = count(counts.get("mappedSourceRecords"));
        unresolvedRecords = count(counts.get("unresolvedSourceRecords")); explicitTargets = count(counts.get("explicitNaifTargets"));
        availableTargets = count(counts.get("availableTargetsAtAuditEpoch"));
        require(mappedRecords <= sourceRecords && unresolvedRecords == sourceRecords - mappedRecords
                && explicitTargets <= mappedRecords && availableTargets <= explicitTargets, "inconsistent coverage counts");
        Map<String, Object> windows = object(value.get("windowCounts"));
        dependencyCovered = count(windows.get("dependencyCoveredTargets")); dependencyGaps = count(windows.get("targetsWithDependencyGaps"));
        require(windows.containsKey("numericallyCertifiedWholeWindowTargets") && windows.get("numericallyCertifiedWholeWindowTargets") == null
                && dependencyCovered <= explicitTargets && dependencyGaps == explicitTargets - dependencyCovered, "invalid coverage window counts");
        Map<String, Object> reasons = object(value.get("unresolvedReasons"));
        require(reasons.size() <= 128, "too many coverage reasons");
        SortedMap<String, Long> parsedReasons = new TreeMap<>();
        long remaining = unresolvedRecords;
        for (Map.Entry<String, Object> entry : reasons.entrySet()) {
            require(entry.getKey().matches("[a-z0-9][a-z0-9-]{0,127}"), "invalid coverage reason");
            long total = count(entry.getValue()); require(total <= remaining, "inconsistent coverage reasons");
            remaining -= total; parsedReasons.put(entry.getKey(), total);
        }
        require(remaining == 0, "incomplete coverage reasons");
        unresolvedReasons = Collections.unmodifiableSortedMap(parsedReasons);
        auditEt = finite(value.get("auditEt"));
        Object rawCoverage = value.get("coverage");
        require(rawCoverage instanceof List && !((List<?>) rawCoverage).isEmpty() && ((List<?>) rawCoverage).size() <= 256, "invalid coverage status buckets");
        List<CoverageBucket> parsedCoverage = new ArrayList<>(); Set<String> coverageKeys = new HashSet<>(); long coverageTotal = 0;
        for (Object entry : (List<?>) rawCoverage) {
            CoverageBucket bucket = new CoverageBucket(object(entry), catalogVersion, auditEt);
            String key = bucket.datasetVersion + "\u0000" + bucket.source + "\u0000" + bucket.model + "\u0000" + Double.toString(bucket.auditEt) + "\u0000" + bucket.frame;
            require(coverageKeys.add(key), "repeated coverage status bucket");
            Long next = sum(coverageTotal, bucket.exact, bucket.approximate, bucket.missing);
            require(next != null, "coverage status bucket total overflow"); coverageTotal = next;
            parsedCoverage.add(bucket);
        }
        require(coverageTotal == sourceRecords, "coverage status buckets do not reconcile");
        coverage = Collections.unmodifiableList(parsedCoverage);
        Map<String, Object> window = object(value.get("requestedWindow"));
        windowStartEt = finite(window.get("startEt")); windowEndEt = finite(window.get("endEt"));
        require(TIME_SCALE.equals(window.get("timeScale")) && windowStartEt <= windowEndEt, "invalid coverage window");
        // The independent audit epoch is not required to lie inside this window.
    }

    public static CoverageReport decode(byte[] summary, byte[] manifest) throws StateTileDecoder.ProtocolException {
        require(summary != null && summary.length > 0 && summary.length <= 64 * 1024, "coverage summary size invalid");
        require(manifest != null && manifest.length > 0 && manifest.length <= 8 * 1024 * 1024, "coverage manifest size invalid");
        return new CoverageReport(object(StateTileDecoder.parseJson(summary)), object(StateTileDecoder.parseJson(manifest)));
    }

    @SuppressWarnings("unchecked") private static Map<String, Object> object(Object value) throws StateTileDecoder.ProtocolException {
        require(value instanceof Map, "coverage object expected"); return (Map<String, Object>) value;
    }
    private static String string(Map<String, Object> value, String key) throws StateTileDecoder.ProtocolException {
        require(value.get(key) instanceof String, "coverage string expected"); return (String) value.get(key);
    }
    private static String text(Object value, boolean optional, int max) throws StateTileDecoder.ProtocolException {
        if (optional && value == null) return "";
        require(value instanceof String, "coverage string expected"); String result = (String) value;
        require((optional || !result.isEmpty()) && result.length() <= max, "coverage string exceeds limit");
        for (int i = 0; i < result.length(); i++) require(result.charAt(i) >= 32 && result.charAt(i) != 127, "coverage string has control character");
        return result;
    }
    private static String hash(Map<String, Object> value, String key) throws StateTileDecoder.ProtocolException {
        String result = string(value, key); require(result.matches("[a-f0-9]{64}"), "coverage hash invalid"); return result;
    }
    private static double finite(Object value) throws StateTileDecoder.ProtocolException {
        require(value instanceof Number, "coverage number expected"); double result = ((Number) value).doubleValue();
        require(Double.isFinite(result), "coverage number must be finite"); return result;
    }
    private static long count(Object value) throws StateTileDecoder.ProtocolException {
        double result = finite(value); require(result >= 0 && result <= 9_007_199_254_740_991d && result == Math.rint(result), "coverage count invalid");
        return (long) result;
    }
    private static Long sum(long... values) {
        long total = 0;
        for (long value : values) { if (value < 0 || Long.MAX_VALUE - total < value || total + value > 9_007_199_254_740_991L) return null; total += value; }
        return total;
    }
    private static void require(boolean valid, String reason) throws StateTileDecoder.ProtocolException {
        if (!valid) throw new StateTileDecoder.ProtocolException(reason);
    }
}
